superbench/benchmarks/micro_benchmarks/directx_render_performance/RenderGeometryPass.cpp

```diff
@@ -0,0 +1,139 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "RenderGeometryPass.h"
+
+int RenderGeometryPass::DefineRootParameters(std::vector<CD3DX12_ROOT_PARAMETER> &rootParameters) {
+    int numRootParams = 5;
+    rootParameters.resize(numRootParams);
+
+    std::unique_ptr<CD3DX12_DESCRIPTOR_RANGE> texTable0 = std::make_unique<CD3DX12_DESCRIPTOR_RANGE>();
+    texTable0->Init(D3D12_DESCRIPTOR_RANGE_TYPE_SRV, 1, 0, 0);
+    std::unique_ptr<CD3DX12_DESCRIPTOR_RANGE> texTable1 = std::make_unique<CD3DX12_DESCRIPTOR_RANGE>();
+    texTable1->Init(D3D12_DESCRIPTOR_RANGE_TYPE_SRV, m_numShaderResource - 1, 1, 0);
+
+    rootParameters[0].InitAsConstantBufferView(0);                                                  // obj cb
+    rootParameters[1].InitAsConstantBufferView(1);                                                  // pass cb
+    rootParameters[2].InitAsConstantBufferView(2);                                                  // material cb
+    rootParameters[3].InitAsDescriptorTable(1, texTable0.release(), D3D12_SHADER_VISIBILITY_PIXEL); // cube texture
+    rootParameters[4].InitAsDescriptorTable(1, texTable1.release(), D3D12_SHADER_VISIBILITY_PIXEL); // texture array
+
+    return numRootParams;
+}
+
+int RenderGeometryPass::DefineStaticSamplers(std::vector<CD3DX12_STATIC_SAMPLER_DESC> &samplers) {
+    int samplersCount = 1;
+    samplers.resize(samplersCount);
+
+    CD3DX12_STATIC_SAMPLER_DESC anisotropicWrap(0,                               // shaderRegister
+                                                D3D12_FILTER_ANISOTROPIC,        // filter
+                                                D3D12_TEXTURE_ADDRESS_MODE_WRAP, // addressU
+                                                D3D12_TEXTURE_ADDRESS_MODE_WRAP, // addressV
+                                                D3D12_TEXTURE_ADDRESS_MODE_WRAP, // addressW
+                                                0.0f,                            // mipLODBias
+                                                8);                              // maxAnisotropy
+    samplers[0] = anisotropicWrap;
+
+    return samplersCount;
+}
+
+void RenderGeometryPass::CreateShaderResourceView(ID3D12Device *device, ID3D12GraphicsCommandList *cmdList, int width,
```

Comment:
>device [](http://example.com/codeflow?start=64&length=6)

do we need to check the pointer before we use it? I know that for this device variable, you have ThrowIfFailed to guard it, but for other pointer parameters, I suggest we check its value before.

Reply:
yes, updated to check the pointer where firstly use it
```
void RenderApp::Initialize() {
    if (m_deviceResources == nullptr) {
        throw std::runtime_error("DeviceResources is nullptr");
    }
    m_deviceResources->SetWindow(m_hMainWnd, m_width, m_height);
    m_deviceResources->CreateDeviceResources();
    CreateDeviceDependentResources();

    m_deviceResources->CreateWindowSizeDependentResources();
    CreateWindowSizeDependentResources();

    // Wait until initialization is complete.
    // Execute the initialization commands.
    m_deviceResources->WaitForGpu();

    auto device = m_deviceResources->GetD3DDevice();
    auto commandQueue = m_deviceResources->GetCommandQueue();
    m_gpuTimer.init(device, commandQueue, m_maxTimerNum, D3D12::QueueType::compute);
    m_outfile.open(m_opts->m_outfile, std::ios_base::out);
}

void RenderApp::CreateDeviceDependentResources() {
    auto device = m_deviceResources->GetD3DDevice();
    if (device == nullptr) {
        throw std::runtime_error("D3D12Device is nullptr");
    }
    // Create a fence for synchronizing between different frames
    ThrowIfFailed(device->CreateFence(m_deviceResources->GetCurrentFrameIndex(), D3D12_FENCE_FLAG_NONE,
                                      IID_PPV_ARGS(m_fence.ReleaseAndGetAddressOf())));

    // Start off the fence with the current frame index
    uint64_t currentIdx = m_deviceResources->GetCurrentFrameIndex();
    m_deviceResources->GetCommandQueue()->Signal(m_fence.Get(), currentIdx);

    CreateRootSignatures(device);
    BuildPipelineStates(device);
}

void RenderApp::CreateWindowSizeDependentResources() {
    auto device = m_deviceResources->GetD3DDevice();
    auto rtvHeap = m_deviceResources->m_rtvDescriptorHeap.Get();
    auto pCmdList = m_deviceResources->GetCommandList();
    auto cmdListAlloc = m_deviceResources->GetCommandAllocator();
    auto cmdQueue = m_deviceResources->GetCommandQueue();
    if (device == nullptr) {
        throw std::runtime_error("D3D12Device is nullptr");
    }
    if (rtvHeap == nullptr) {
        throw std::runtime_error("RTVDescriptorHeap is nullptr");
    }
    if (pCmdList == nullptr) {
        throw std::runtime_error("CommandList is nullptr");
    }
    if (cmdListAlloc == nullptr) {
        throw std::runtime_error("CommandAllocator is nullptr");
    }
    if (cmdQueue == nullptr) {
        throw std::runtime_error("CommandQueue is nullptr");
    }

    ThrowIfFailed(cmdListAlloc->Reset());
    ThrowIfFailed(pCmdList->Reset(cmdListAlloc, nullptr));
```

Reply:
well, this is good, but my suggestion is that we should check the pointer here instead of the caller. We can make sure that when RenderApp calls it, it works well, but when you or some others writes another application, the pointer checking may be forgotten again. Checking the pointer's validity is the function's responsibility.